### ISSUE TITLE
 can't pickle _ElementUnicodeResult objects

### ISSUE THREAD

**Issue by gpopovic** [GoogleCloudPlatform/DataflowPythonSDK#29]
I just can't figure out why this is happening ...
I'm returning {} filled with data on a beam.Map() function.

Everything works on local machine but i keep getting this error when i'm running in cloud.

`n exception was raised when trying to execute the work item <BatchWorkItem s01 steps=read crawl+process+save/native_write 660446402423394471> : Traceback (most recent call last): File "/usr/local/lib/python2.7/dist-packages/dataflow_worker/batchworker.py", line 462, in do_work work_executor.execute() File "dataflow_worker/executor.py", line 889, in dataflow_worker.executor.MapTaskExecutor.execute (dataflow_worker/executor.c:23689) op.start() File "dataflow_worker/executor.py", line 184, in dataflow_worker.executor.ReadOperation.start (dataflow_worker/executor.c:6766) def start(self): File "dataflow_worker/executor.py", line 189, in dataflow_worker.executor.ReadOperation.start (dataflow_worker/executor.c:6673) with self.spec.source.reader() as reader: File "dataflow_worker/executor.py", line 199, in dataflow_worker.executor.ReadOperation.start (dataflow_worker/executor.c:6618) self.output(windowed_value) File "dataflow_worker/executor.py", line 132, in dataflow_worker.executor.Operation.output (dataflow_worker/executor.c:5093) cython.cast(Receiver, self.receivers[output_index]).receive(windowed_value) File "dataflow_worker/executor.py", line 79, in dataflow_worker.executor.ConsumerSet.receive (dataflow_worker/executor.c:3331) cython.cast(Operation, consumer).process(windowed_value) File "dataflow_worker/executor.py", line 508, in dataflow_worker.executor.DoOperation.process (dataflow_worker/executor.c:14242) self.dofn_receiver.receive(o) File "apache_beam/runners/common.py", line 101, in apache_beam.runners.common.DoFnRunner.receive (apache_beam/runners/common.c:3344) self.process(windowed_value) File "apache_beam/runners/common.py", line 129, in apache_beam.runners.common.DoFnRunner.process (apache_beam/runners/common.c:4307) self.reraise_augmented(exn) File "apache_beam/runners/common.py", line 140, in apache_beam.runners.common.DoFnRunner.reraise_augmented (apache_beam/runners/common.c:4670) raise type(exn), args, sys.exc_info()[2] File "apache_beam/runners/common.py", line 127, in apache_beam.runners.common.DoFnRunner.process (apache_beam/runners/common.c:4245) self._process_outputs(element, self.dofn_process(self.context)) File "apache_beam/runners/common.py", line 183, in apache_beam.runners.common.DoFnRunner._process_outputs (apache_beam/runners/common.c:5424) self.main_receivers.receive(windowed_value) File "dataflow_worker/executor.py", line 77, in dataflow_worker.executor.ConsumerSet.receive (dataflow_worker/executor.c:3297) self.update_counters_start(windowed_value) File "dataflow_worker/executor.py", line 83, in dataflow_worker.executor.ConsumerSet.update_counters_start (dataflow_worker/executor.c:3496) self.opcounter.update_from(windowed_value) File "dataflow_worker/opcounters.py", line 57, in dataflow_worker.opcounters.OperationCounters.update_from (dataflow_worker/opcounters.c:2131) self.do_sample(windowed_value) File "dataflow_worker/opcounters.py", line 71, in dataflow_worker.opcounters.OperationCounters.do_sample (dataflow_worker/opcounters.c:2685) self.coder_impl.get_estimated_size_and_observables(windowed_value)) File "apache_beam/coders/coder_impl.py", line 511, in apache_beam.coders.coder_impl.WindowedValueCoderImpl.get_estimated_size_and_observables (apache_beam/coders/coder_impl.c:16793) def get_estimated_size_and_observables(self, value, nested=False): File "apache_beam/coders/coder_impl.py", line 520, in apache_beam.coders.coder_impl.WindowedValueCoderImpl.get_estimated_size_and_observables (apache_beam/coders/coder_impl.c:16525) c.get_estimated_size_and_observables(value.value, nested=True)) File "apache_beam/coders/coder_impl.py", line 100, in apache_beam.coders.coder_impl.CoderImpl.get_estimated_size_and_observables (apache_beam/coders/coder_impl.c:3271) return self.estimate_size(value, nested), [] File "apache_beam/coders/coder_impl.py", line 129, in apache_beam.coders.coder_impl.StreamCoderImpl.estimate_size (apache_beam/coders/coder_impl.c:4279) self.encode_to_stream(value, out, nested) File "apache_beam/coders/coder_impl.py", line 243, in apache_beam.coders.coder_impl.FastPrimitivesCoderImpl.encode_to_stream (apache_beam/coders/coder_impl.c:7375) self.encode_to_stream(v, stream, True) File "apache_beam/coders/coder_impl.py", line 236, in apache_beam.coders.coder_impl.FastPrimitivesCoderImpl.encode_to_stream (apache_beam/coders/coder_impl.c:7258) self.encode_to_stream(e, stream, True) File "apache_beam/coders/coder_impl.py", line 246, in apache_beam.coders.coder_impl.FastPrimitivesCoderImpl.encode_to_stream (apache_beam/coders/coder_impl.c:7410) self.fallback_coder_impl.encode_to_stream(value, stream, nested) File "apache_beam/coders/coder_impl.py", line 149, in apache_beam.coders.coder_impl.CallbackCoderImpl.encode_to_stream (apache_beam/coders/coder_impl.c:4749) return stream.write(self._encoder(value), nested) File "/usr/lib/python2.7/copy_reg.py", line 70, in _reduce_ex raise TypeError, "can't pickle %s objects" % base.__name__ TypeError: can't pickle _ElementUnicodeResult objects [while running 'process']`

I have absolutely no idea how to debug this..

**Comment by aaltay** [comment]
In general the elements of a PCollection needs to be codeable. The default coder is PickleCoder and it seems like _ElementUnicodeResult not picklable. This code path is not exercised in local runs and that is why it is working locally. (As a side note, local runner would verify these things in the future.)

What is the type of elements of your PCollections? It is possible that _ElementUnicodeResult is coming as a nested element of something else. Would it be possible for you to use some other element type instead?

**Comment by gpopovic** [comment]
@aaltay it's a dictionary ..something like that
{
      "a":"b",
      "c": ["a","b","c"...]
}

**Comment by gpopovic** [comment]
@aaltay I managed to reproduce error on local run, with using cPickle..

Sometimes lxml returns _ElementUnicodeResult instead of unicode/string... when compared to string/unicode it will be equal, but will fail when it needs to be decoded/encoded.

After changing type from _ElementUnicodeResult to unicode..everything works like a charm..
I spent whole day trying to figure out why this doesn't work..and it was so simple.

Verifying if object can be coded/decoded on local runner would help fixing bugs like this.
